File: sdk/rust/src/transaction.rs
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;

use async_trait::async_trait;
use hedera_proto::services;
use prost::Message;
use sha2::{Digest, Sha384};
use time::Duration;
use tonic::transport::Channel;
use tonic::{Response, Status};

use crate::execute::{execute, Execute};
use crate::{AccountId, Client, Error, Signer, ToProtobuf, TransactionId, TransactionResponse};

pub struct Transaction<D> {
    pub(crate) data: D,
    node_account_ids: Option<Vec<AccountId>>,
    transaction_valid_duration: Duration,
    max_transaction_fee: Option<u64>,
    transaction_memo: String,
    signers: Vec<Box<dyn Signer>>,
    // TODO: payer_account_id: Option<AccountId>
    // TODO: transaction_id: Option<TransactionId>
    //  While transaction IDs are generated on-demand by using the operator, if you don't have
    //  an operator configured, you are going to have a bad day
}

impl<D> Transaction<D>
where
    D: Default,
{
    fn default() -> Self {
        Self {
            data: D::default(),
            node_account_ids: None,
            transaction_valid_duration: Duration::seconds(120),
            transaction_memo: String::new(),
            max_transaction_fee: None,
            signers: Vec::new(),
        }
    }
}

impl<D> Transaction<D>
where
    D: Default,
{
    #[inline]
    pub fn new() -> Self {
        Self::default()
    }
}

impl<D> Transaction<D> {
    /// Set the account IDs of the nodes that this transaction may be submitted to.
    ///
    /// Defaults to the full list of nodes configured on the client.
    ///
    pub fn node_account_ids(&mut self, ids: impl IntoIterator<Item = AccountId>) -> &mut Self {
        self.node_account_ids = Some(ids.into_iter().collect());
        self
    }

    /// Sets the duration that this transaction is valid for, once finalized and signed.
    ///
    /// Defaults to 120 seconds (or two minutes).
    ///  
    pub fn transaction_valid_duration(&mut self, duration: Duration) -> &mut Self {
        self.transaction_valid_duration = duration;
        self
    }

    /// Set the maximum transaction fee the operator (paying account) is willing to pay.
    pub fn max_transaction_fee(&mut self, fee: u64) -> &mut Self {
        self.max_transaction_fee = Some(fee);
        self
    }

    /// Set a note or description that should be recorded in the transaction record (maximum length
    /// of 100 characters).
    pub fn transaction_memo(&mut self, memo: impl AsRef<str>) -> &mut Self {
        self.transaction_memo = memo.as_ref().to_owned();
        self
    }

    /// Adds the signer to the list of signers that will sign this transaction before sending
    /// to the network.
    pub fn signer<S>(&mut self, signer: &S) -> &mut Self
    where
        S: Signer + Clone,
    {
        self.signers.push(Box::new(signer.clone()));
        self
    }
}

#[async_trait]
pub trait TransactionExecute {
    fn default_max_transaction_fee() -> u64 {
        2 * 100_000_000 // 2 hbar
    }

    async fn execute(
        channel: Channel,
        request: services::Transaction,
    ) -> Result<tonic::Response<services::TransactionResponse>, tonic::Status>;
}

impl<D> Transaction<D>
where
    D: ToProtobuf<Protobuf = services::transaction_body::Data>,
    Self: TransactionExecute,
{
    #[allow(deprecated)]
    fn to_transaction_body_protobuf(
        &self,
        node_account_id: AccountId,
        transaction_id: TransactionId,
        client_max_transaction_fee: &Arc<AtomicU64>,
    ) -> services::TransactionBody {
        let data = self.data.to_protobuf();

        let max_transaction_fee = self.max_transaction_fee.unwrap_or_else(|| {
            // no max has been set on the *transaction*
            // check if there is a global max set on the client
            match client_max_transaction_fee.load(Ordering::Relaxed) {
                max if max > 1 => max,

                // no max has been set on the client either
                // fallback to the hard-coded default for this transaction type
                _ => Self::default_max_transaction_fee(),
            }
        });

        services::TransactionBody {
            data: Some(data),
            transaction_id: Some(transaction_id.to_protobuf()),
            transaction_valid_duration: Some(self.transaction_valid_duration.into()),
            memo: self.transaction_memo.clone(),
            node_account_id: Some(node_account_id.to_protobuf()),
            generate_record: false,
            transaction_fee: max_transaction_fee,
        }
    }
}

#[async_trait]
impl<D> Execute for Transaction<D>
where
    D: ToProtobuf<Protobuf = services::transaction_body::Data>,
    Self: TransactionExecute,
{
    type GrpcRequest = services::Transaction;

    type GrpcResponse = services::TransactionResponse;

    type RequestContext = ();

    type ResponseContext = [u8; 48];

    type Response = TransactionResponse;

    fn node_account_ids(&self) -> Option<&[AccountId]> {
        self.node_account_ids.as_deref()
    }

    fn transaction_id(&self) -> Option<TransactionId> {
        // TODO: explicit
        None
    }

    fn requires_transaction_id() -> bool {
        true
    }

    async fn make_request(
        &self,
        client: &Client,
        transaction_id: Option<TransactionId>,
        node_account_id: AccountId,
        _context: &Self::RequestContext,
    ) -> crate::Result<(Self::GrpcRequest, Self::ResponseContext)> {
        let transaction_id = transaction_id.unwrap();

        let transaction_body = self.to_transaction_body_protobuf(
            node_account_id,
            transaction_id.clone(),
            &client.max_transaction_fee,
        );

        let body_bytes = transaction_body.encode_to_vec();

        let mut signatures = Vec::with_capacity(self.signers.len());

        for signer in &self.signers {
            // TODO: should we run the signers in parallel?
            let signature = signer.sign(&body_bytes).await.map_err(Error::signature)?;

            signatures.push(signature.to_protobuf());
        }

        let signed_transaction = services::SignedTransaction {
            body_bytes,
            sig_map: Some(services::SignatureMap { sig_pair: signatures }),
        };

        let signed_transaction_bytes = signed_transaction.encode_to_vec();

        let transaction_hash = Sha384::digest(&signed_transaction_bytes);

        let transaction =
            services::Transaction { signed_transaction_bytes, ..services::Transaction::default() };

        Ok((transaction, transaction_hash.into()))
    }

    async fn execute(
        channel: Channel,
        request: Self::GrpcRequest,
    ) -> Result<Response<Self::GrpcResponse>, Status> {
        <Self as TransactionExecute>::execute(channel, request).await
    }

    fn make_response(
        _response: Self::GrpcResponse,
        transaction_hash: Self::ResponseContext,
        node_account_id: AccountId,
        transaction_id: Option<TransactionId>,
    ) -> crate::Result<Self::Response> {
        Ok(TransactionResponse {
            node_account_id,
            transaction_id: transaction_id.unwrap(),
            transaction_hash,
        })
    }

    fn response_pre_check_status(response: &Self::GrpcResponse) -> crate::Result<i32> {
        Ok(response.node_transaction_precheck_code)
    }
}

impl<D> Transaction<D>
where
    D: ToProtobuf<Protobuf = services::transaction_body::Data>,
    Self: TransactionExecute,
{
    pub async fn execute(&self, client: &Client) -> crate::Result<TransactionResponse> {
        execute(client, self, ()).await
    }
}

File: sdk/swift/Sources/Hedera/Transaction.swift
/*
 * ‌
 * Hedera Swift SDK
 * ​
 * Copyright (C) 2022 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import Foundation
import GRPC
import HederaProtobufs

/// A transaction that can be executed on the Hedera network.
public class Transaction: ValidateChecksums {
    public typealias Response = TransactionResponse

    public init() {}

    internal init(protobuf proto: Proto_TransactionBody) throws {
        transactionValidDuration = .fromProtobuf(proto.transactionValidDuration)
        maxTransactionFee = .fromTinybars(Int64(proto.transactionFee))
        transactionMemo = proto.memo
        transactionId = try .fromProtobuf(proto.transactionID)
    }

    internal private(set) final var signers: [Signer] = []
    internal private(set) final var sources: TransactionSources?
    public private(set) final var isFrozen: Bool = false

    private final var `operator`: Operator?

    internal var defaultMaxTransactionFee: Hbar {
        2
    }

    internal func toTransactionDataProtobuf(_ chunkInfo: ChunkInfo) -> Proto_TransactionBody.OneOf_Data {
        fatalError("Method `Transaction.toTransactionDataProtobuf` must be overridden by `\(type(of: self))`")
    }

    internal func transactionExecute(_ channel: GRPCChannel, _ request: Proto_Transaction) async throws
        -> Proto_TransactionResponse
    {
        fatalError("Method `Transaction.transactionExecute` must be overridden by `\(type(of: self))`")
    }

    public final var nodeAccountIds: [AccountId]? {
        willSet {
            ensureNotFrozen(fieldName: "nodeAccountIds")
        }
    }

    @discardableResult
    public func nodeAccountIds(_ nodeAccountIds: [AccountId]) -> Self {
        self.nodeAccountIds = nodeAccountIds

        return self
    }

    /// Explicit transaction ID for this transaction.
    public final var transactionId: TransactionId? {
        willSet {
            ensureNotFrozen(fieldName: "transactionId")
        }
    }

    /// The maximum allowed transaction fee for this transaction.
    public final var maxTransactionFee: Hbar? {
        willSet {
            ensureNotFrozen(fieldName: "maxTransactionFee")
        }
    }

    /// Sets the maximum allowed transaction fee for this transaction.
    @discardableResult
    public final func maxTransactionFee(_ maxTransactionFee: Hbar) -> Self {
        self.maxTransactionFee = maxTransactionFee

        return self
    }

    public final var transactionValidDuration: Duration? {
        willSet {
            ensureNotFrozen(fieldName: "transactionValidDuration")
        }
    }

    @discardableResult
    public final func transactionValidDuration(_ transactionValidDuration: Duration) -> Self {
        self.transactionValidDuration = transactionValidDuration

        return self
    }

    public final var transactionMemo: String = "" {
        willSet {
            ensureNotFrozen(fieldName: "transactionMemo")
        }
    }

    @discardableResult
    public final func transactionMemo(_ transactionMemo: String) -> Self {
        self.transactionMemo = transactionMemo

        return self
    }

    /// Sets the explicit transaction ID for this transaction.
    @discardableResult
    public final func transactionId(_ transactionId: TransactionId) -> Self {
        self.transactionId = transactionId

        return self
    }

    internal func addSignatureSigner(_ signer: Signer) {
        precondition(isFrozen)

        precondition(nodeAccountIds?.count == 1, "cannot manually add a signature to a transaction with multiple nodes")

        // swiftlint:disable:next force_try
        let sources = try! makeSources()

        self.sources = sources.signWithSigners([signer])
    }

    @discardableResult
    public final func sign(_ privateKey: PrivateKey) -> Self {
        self.signWithSigner(.privateKey(privateKey))

        return self
    }

    @discardableResult
    public final func signWith(_ publicKey: PublicKey, _ signer: @escaping (Data) -> (Data)) -> Self {
        self.signWithSigner(Signer(publicKey, signer))

        return self
    }

    internal final func signWithSigner(_ signer: Signer) {
        self.signers.append(signer)
    }

    @discardableResult
    public final func freeze() throws -> Self {
        try freezeWith(nil)
    }

    @discardableResult
    public final func freezeWith(_ client: Client?) throws -> Self {
        if isFrozen {
            return self
        }

        guard let nodeAccountIds = self.nodeAccountIds ?? client?.randomNodeIds() else {
            throw HError(
                kind: .freezeUnsetNodeAccountIds, description: "transaction frozen without client or explicit node IDs")
        }

        let maxTransactionFee = self.maxTransactionFee ?? client?.maxTransactionFee

        let `operator` = client?.operator

        self.nodeAccountIds = nodeAccountIds
        self.maxTransactionFee = maxTransactionFee
        self.`operator` = `operator`

        isFrozen = true

        if client?.isAutoValidateChecksumsEnabled() == true {
            try validateChecksums(on: client!)
        }

        return self
    }

    @discardableResult
    internal final func makeSources() throws -> TransactionSources {
        precondition(isFrozen)
        if let sources = sources {
            return sources.signWithSigners(self.signers)
        }

        let transactions = try self.makeTransactionList()

        // swiftlint:disable:next force_try
        return try! TransactionSources(transactions: transactions)
    }

    public func execute(_ client: Client, _ timeout: TimeInterval? = nil) async throws -> Response {
        try freezeWith(client)

        if let sources = sources {
            return try await SourceTransaction(inner: self, sources: sources).execute(client, timeout: timeout)
        }

        return try await executeAny(client, self, timeout)
    }

    public static func fromBytes(_ bytes: Data) throws -> Transaction {
        let list: [Proto_Transaction]
        do {
            let tmp = try Proto_TransactionList(contiguousBytes: bytes)

            if tmp.transactionList.isEmpty {
                list = [try Proto_Transaction(contiguousBytes: bytes)]
            } else {
                list = tmp.transactionList
            }
        } catch {
            throw HError.fromProtobuf(String(describing: error))
        }

        let sources = try TransactionSources(transactions: list)

        let transactionBodies = try sources.signedTransactions.map { signed -> Proto_TransactionBody in
            do {
                return try Proto_TransactionBody(contiguousBytes: signed.bodyBytes)
            } catch {
                throw HError.fromProtobuf(String(describing: error))
            }
        }

        do {
            let (first, rest) = (transactionBodies[0], transactionBodies[1...])

            for body in rest {
                guard protoTransactionBodyEqual(body, first) else {
                    throw HError.fromProtobuf("transaction parts unexpectedly unequal")
                }
            }

        }

        let transactionData =
            try sources
            .chunks
            .compactMap { $0.signedTransactions.first }
            .lazy
            .map { signed -> Proto_TransactionBody in
                do {
                    return try Proto_TransactionBody(contiguousBytes: signed.bodyBytes)
                } catch {
                    throw HError.fromProtobuf(String(describing: error))
                }

            }
            .map { body -> Proto_TransactionBody.OneOf_Data in
                guard let data = body.data else {
                    throw HError.fromProtobuf("Unexpected missing `data`")
                }

                return data
            }

        // note: this creates the transaction in a unfrozen state by pure need.
        let transaction = try AnyTransaction.fromProtobuf(transactionBodies[0], transactionData).transaction

        transaction.nodeAccountIds = sources.nodeAccountIds
        transaction.sources = sources
        // explicitly avoid `freeze`.
        transaction.isFrozen = true

        return transaction
    }

    public final func toBytes() throws -> Data {
        precondition(isFrozen, "Transaction must be frozen to call `toBytes`")

        if let sources = self.sources {
            return sources.toBytes()
        }

        let transactionList = try Proto_TransactionList.with { proto in
            proto.transactionList = try self.makeTransactionList()
        }

        // swiftlint:disable:next force_try
        return try! transactionList.serializedData()
    }

    internal func validateChecksums(on ledgerId: LedgerId) throws {
        try nodeAccountIds?.validateChecksums(on: ledgerId)
        try transactionId?.validateChecksums(on: ledgerId)
    }

    internal final func ensureNotFrozen(fieldName: String? = nil) {
        if let fieldName = fieldName {
            precondition(!isFrozen, "\(fieldName) cannot be set while `\(type(of: self))` is frozen")
        } else {
            precondition(
                !isFrozen,
                "`\(type(of: self))` is immutable; it has at least one signature or has been explicitly frozen")
        }
    }
}

extension Transaction {
    internal final func makeResponse(
        _: Proto_TransactionResponse, _ context: TransactionHash, _ nodeAccountId: AccountId,
        _ transactionId: TransactionId?
    ) -> Response {
        TransactionResponse(nodeAccountId: nodeAccountId, transactionId: transactionId!, transactionHash: context)
    }

    internal final func makeErrorPrecheck(_ status: Status, _ transactionId: TransactionId?) -> HError {
        guard let transactionId = transactionId else {
            return HError(
                kind: .transactionNoIdPreCheckStatus(status: status),
                description: "transaction without transaction id failed pre-check with status `\(status)`"
            )
        }
        return HError(
            kind: .transactionPreCheckStatus(status: status, transactionId: transactionId),
            description: "transaction `\(transactionId)` failed pre-check with status `\(status)"
        )
    }

    internal static func responsePrecheckStatus(_ response: Proto_TransactionResponse) -> Int32 {
        Int32(response.nodeTransactionPrecheckCode.rawValue)
    }
}

extension Transaction {
    fileprivate func makeTransactionList() throws -> [Proto_Transaction] {
        assert(self.isFrozen)

        // todo: fix this with chunked transactions.
        guard let initialTransactionId = self.transactionId ?? self.operator?.generateTransactionId() else {
            throw HError.noPayerAccountOrTransactionId
        }

        let usedChunks = (self as? ChunkedTransaction)?.usedChunks ?? 1
        let nodeAccountIds = nodeAccountIds!

        var transactionList: [Proto_Transaction] = []

        // Note: This ordering is *important*,
        // there's no documentation for it but `TransactionList` is sorted by chunk number,
        // then `node_id` (in the order they were added to the transaction)
        for chunk in 0..<usedChunks {
            let currentTransactionId: TransactionId
            switch chunk {
            case 0:
                currentTransactionId = initialTransactionId
            default:
                guard let `operator` = self.operator else {
                    throw HError.noPayerAccountOrTransactionId
                }

                currentTransactionId = `operator`.generateTransactionId()
            }

            for nodeAccountId in nodeAccountIds {
                let chunkInfo = ChunkInfo(
                    current: chunk,
                    total: usedChunks,
                    initialTransactionId: initialTransactionId,
                    currentTransactionId: currentTransactionId,
                    nodeAccountId: nodeAccountId
                )

                transactionList.append(self.makeRequestInner(chunkInfo: chunkInfo).0)
            }
        }

        return transactionList
    }

    internal func makeRequestInner(chunkInfo: ChunkInfo) -> (Proto_Transaction, TransactionHash) {
        assert(self.isFrozen)

        let body: Proto_TransactionBody = self.toTransactionBodyProtobuf(chunkInfo)

        // swiftlint:disable:next force_try
        let bodyBytes = try! body.serializedData()

        var signatures: [SignaturePair] = []

        if let `operator` = self.operator {
            // todo: avoid the `.map(xyz).collect()`
            let operatorSignature = `operator`.signer.sign(bodyBytes)

            signatures.append(SignaturePair((`operator`.signer.publicKey, operatorSignature)))
        }

        for signer in self.signers where signatures.allSatisfy({ $0.publicKey != signer.publicKey }) {
            let signature = signer(bodyBytes)
            signatures.append(SignaturePair(signature))
        }

        let signedTransaction = Proto_SignedTransaction.with { proto in
            proto.bodyBytes = bodyBytes
            proto.sigMap.sigPair = signatures.toProtobuf()
        }

        // swiftlint:disable:next force_try
        let signedTransactionBytes = try! signedTransaction.serializedData()

        let transactionHash = TransactionHash(hashing: signedTransactionBytes)

        let transaction = Proto_Transaction.with { $0.signedTransactionBytes = signedTransactionBytes }

        return (transaction, transactionHash)
    }

    private func toTransactionBodyProtobuf(_ chunkInfo: ChunkInfo) -> Proto_TransactionBody {
        assert(isFrozen)
        let data = toTransactionDataProtobuf(chunkInfo)

        let maxTransactionFee = self.maxTransactionFee ?? self.defaultMaxTransactionFee

        return .with { proto in
            proto.data = data
            proto.transactionID = chunkInfo.currentTransactionId.toProtobuf()
            proto.transactionValidDuration = (self.transactionValidDuration ?? .minutes(2)).toProtobuf()
            proto.memo = self.transactionMemo
            proto.nodeAccountID = chunkInfo.nodeAccountId.toProtobuf()
            proto.generateRecord = false
            proto.transactionFee = UInt64(maxTransactionFee.toTinybars())
        }
    }
}

extension Transaction: Execute {
    internal typealias GrpcRequest = Proto_Transaction
    internal typealias GrpcResponse = Proto_TransactionResponse
    internal typealias Context = TransactionHash

    internal var explicitTransactionId: TransactionId? {
        transactionId
    }

    internal var requiresTransactionId: Bool { true }

    internal func makeRequest(_ transactionId: TransactionId?, _ nodeAccountId: AccountId) throws -> (
        GrpcRequest, TransactionHash
    ) {
        assert(isFrozen)

        guard let transactionId = transactionId else {
            throw HError.noPayerAccountOrTransactionId
        }

        return self.makeRequestInner(chunkInfo: .single(transactionId: transactionId, nodeAccountId: nodeAccountId))
    }

    internal func execute(_ channel: any GRPCChannel, _ request: GrpcRequest) async throws -> GrpcResponse {
        try await transactionExecute(channel, request)
    }
}

/// Returns true if `lhs == rhs`` other than `transactionId` and `nodeAccountId`, false otherwise.
private func protoTransactionBodyEqual(_ lhs: Proto_TransactionBody, _ rhs: Proto_TransactionBody) -> Bool {
    guard lhs.transactionFee == rhs.transactionFee else {
        return false
    }

    guard lhs.transactionValidDuration == rhs.transactionValidDuration else {
        return false
    }

    guard lhs.generateRecord == rhs.generateRecord else {
        return false
    }

    guard lhs.memo == rhs.memo else {
        return false
    }

    let lhsData: Proto_TransactionBody.OneOf_Data
    let rhsData: Proto_TransactionBody.OneOf_Data

    switch (lhs.data, rhs.data) {
    case (nil, nil):
        return true

    case (.some, nil), (nil, .some):
        return false

    case (.some(let lhs), .some(let rhs)):
        lhsData = lhs
        rhsData = rhs
    }

    switch (lhsData, rhsData) {
    case (.consensusSubmitMessage(let lhs), .consensusSubmitMessage(let rhs)):
        guard lhs.hasTopicID == rhs.hasTopicID else {
            return false
        }

        guard lhs.topicID == rhs.topicID else {
            return false
        }

        guard lhs.hasChunkInfo == rhs.hasChunkInfo else {
            return false
        }

        if lhs.hasChunkInfo /* && rhs.hasChunkInfo */ {
            guard lhs.chunkInfo.initialTransactionID == rhs.chunkInfo.initialTransactionID else {
                return false
            }

            guard lhs.chunkInfo.total == rhs.chunkInfo.total else {
                return false
            }

            guard lhs.unknownFields == rhs.unknownFields else {
                return false
            }

            // allow lhs.chunkInfo.number != rhs.chunkInfo.number
        }

    // allow lhs.message != rhs.message

    case (.fileAppend(let lhs), .fileAppend(let rhs)):
        guard lhs.fileID == rhs.fileID else {
            return false
        }

        // allow lhs.contents != rhs.contents

        guard lhs.unknownFields == rhs.unknownFields else {
            return false
        }

    default:
        guard lhsData == rhsData else {
            return false
        }
    }

    return true
}
